src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/CompositeValueProvider.cs

```diff
@@ -160,13 +160,34 @@ public IValueProvider Filter(BindingSource bindingSource)
                 throw new ArgumentNullException(nameof(bindingSource));
             }
 
+            var shouldFilter = false;
+            for (var i = 0; i < Count; i++)
+            {
+                var valueProvider = Items[i];
+                if (valueProvider is IBindingSourceValueProvider)
+                {
+                    shouldFilter = true;
+                    break;
+                }
+            }
+
+            if (!shouldFilter)
+            {
+                // No inner IBindingSourceValueProvider implementations. Result will be empty.
+                return null;
```

Comment:
I'd suggest a sentinel `static readonly Empty` property instead.

Reply:
That would change the observable behaviour of this method. It returns `null` if it can't find a value provider for the binding source. Would also make the `CustomValueProvider` inconsistent with all other `IBindingSourceValueProvider` implementations.

If you feel strongly about this, could return an empty `CompositeValueProvider` in the **new** method. My concern there is inventing a new set of behaviours (can't quite call them concepts), making `IKeyRewriterValueProvider` more different from `IBindingSourceValueProvider `.

That said, I'm ambivalent about the `return null` choice because the only non-test user of this method does the `?? new CompositeValueProvider()` thing. Wouldn't do this if we were starting from scratch. Suggest we change the **caller** to use a `static CompositeValueProvider` field of course. That is, leave `return null` here but handle it (and the non-`CompositeValueProvider` case) better in `DefaultModelBindingContext`. Sound good?